File: api/schemas/session.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from models.session import SessionStatus, FeedbackSource


class AnswerIn(BaseModel):
    question_id: int
    selected_option_id: Optional[int] = None
    text_answer: Optional[str] = None


class SubmitSessionIn(BaseModel):
    answers: List[AnswerIn]


class AnswerOut(BaseModel):
    id: int
    question_id: int
    selected_option_id: Optional[int]
    text_answer: Optional[str]
    is_correct: Optional[bool]
    marks_awarded: float
    auto_marked: bool
    model_config = {"from_attributes": True}


class ScoreOut(BaseModel):
    total_marks_awarded: float
    total_marks_possible: int
    percentage: float
    pass_fail: Optional[bool]
    model_config = {"from_attributes": True}


class FeedbackOut(BaseModel):
    id: int
    generated_by: FeedbackSource
    content: str
    strength_areas: Optional[str]
    weakness_areas: Optional[str]
    recommendations: Optional[str]
    created_at: datetime
    model_config = {"from_attributes": True}


class SessionOut(BaseModel):
    id: int
    assessment_id: int
    learner_id: int
    started_at: datetime
    completed_at: Optional[datetime]
    status: SessionStatus
    answers: List[AnswerOut] = []
    score: Optional[ScoreOut] = None
    feedbacks: List[FeedbackOut] = []
    model_config = {"from_attributes": True}


class SessionStartOut(BaseModel):
    session_id: int
    assessment_id: int
    started_at: datetime
    model_config = {"from_attributes": True}


class TutorFeedbackIn(BaseModel):
    content: str
    strength_areas: Optional[str] = None
    weakness_areas: Optional[str] = None
    recommendations: Optional[str] = None


class MarkAnswerIn(BaseModel):
    marks_awarded: float
    is_correct: bool


class SessionSummaryOut(BaseModel):
    id: int
    assessment_id: int
    assessment_title: str
    learner_id: int
    learner_name: str
    learner_phone: str
    status: SessionStatus
    started_at: datetime
    completed_at: Optional[datetime]
    score: Optional[ScoreOut] = None

File: tutor-desktop/Models/Session.cs
namespace SkyLearnIQ.Tutor.Models;

public class AnswerOut
{
    public int     Id               { get; set; }
    public int     QuestionId       { get; set; }
    public int?    SelectedOptionId { get; set; }
    public string? TextAnswer       { get; set; }
    public bool?   IsCorrect        { get; set; }
    public float   MarksAwarded     { get; set; }
    public bool    AutoMarked       { get; set; }

    public string CorrectDisplay => IsCorrect == true ? "✓ Correct"
                                  : IsCorrect == false ? "✗ Wrong"
                                  : "⏳ Pending";
}

public class ScoreOut
{
    public float  TotalMarksAwarded  { get; set; }
    public int    TotalMarksPossible { get; set; }
    public float  Percentage         { get; set; }
    public bool?  PassFail           { get; set; }

    public string PercentageDisplay => $"{Math.Round(Percentage)}%";
    public string PassFailDisplay   => PassFail == true ? "PASS" : PassFail == false ? "FAIL" : "Pending";
}

public class FeedbackOut
{
    public int      Id              { get; set; }
    public string   GeneratedBy     { get; set; } = "";
    public string   Content         { get; set; } = "";
    public string?  StrengthAreas   { get; set; }
    public string?  WeaknessAreas   { get; set; }
    public string?  Recommendations { get; set; }
    public DateTime CreatedAt       { get; set; }

    public string SourceDisplay => GeneratedBy == "AI" ? "🧠 AI Feedback" : "👨‍🏫 Tutor Feedback";
}

public class SessionSummaryOut
{
    public int       Id               { get; set; }
    public int       AssessmentId     { get; set; }
    public string    AssessmentTitle  { get; set; } = "";
    public int       LearnerId        { get; set; }
    public string    LearnerName      { get; set; } = "";
    public string    LearnerPhone     { get; set; } = "";
    public string    Status           { get; set; } = "";
    public DateTime  StartedAt        { get; set; }
    public DateTime? CompletedAt      { get; set; }
    public ScoreOut? Score            { get; set; }

    public string StatusDisplay  => Status switch
    {
        "COMPLETED"   => "✓ Completed",
        "IN_PROGRESS" => "⏳ In Progress",
        "ABANDONED"   => "✗ Abandoned",
        _             => Status
    };
    public string ScoreDisplay => Score != null ? $"{Math.Round(Score.Percentage)}%" : "—";
    public string DateDisplay  => (CompletedAt ?? StartedAt).ToString("dd MMM yyyy");
}

public class SessionOut
{
    public int              Id           { get; set; }
    public int              AssessmentId { get; set; }
    public int              LearnerId    { get; set; }
    public string           Status       { get; set; } = "";
    public DateTime         StartedAt    { get; set; }
    public DateTime?        CompletedAt  { get; set; }
    public List<AnswerOut>  Answers      { get; set; } = [];
    public ScoreOut?        Score        { get; set; }
    public List<FeedbackOut> Feedbacks   { get; set; } = [];
}
